File: java/test/_regress_linear.java
import java.util.Arrays;
import liba.regress_linear;

class _regress_linear {
    public static void main(String[] args) {
        double x[] = { 0, 1, 2, 3, 4 };
        double y[] = { 1, 2, 3, 4, 5 };
        double r[] = { 0, 0, 0, 0, 0 };
        double coef[] = { 0 };
        regress_linear obj = new regress_linear(coef, 0);
        obj.set_coef(obj.coef());
        obj.set_bias(obj.bias());
        obj.gd(coef, 0, 0);

        obj.zero();
        for (int i = 0; i < 100; ++i) {
            obj.sgd(x, y, 0.1);
        }
        for (int i = 0; i < r.length; i++) {
            double[] val = { x[i] };
            r[i] = obj.eval(val);
        }
        System.out.println(Arrays.toString(r));
        System.out.println(Arrays.toString(obj.err(x, y)));

        obj.zero();
        for (int i = 0; i < 100; ++i) {
            obj.bgd(x, y, 0.1);
        }
        for (int i = 0; i < r.length; i++) {
            double[] val = { x[i] };
            r[i] = obj.eval(val);
        }
        System.out.println(Arrays.toString(r));
        System.out.println(Arrays.toString(obj.err(x, y)));

        obj.zero();
        obj.mgd(x, y, 0.001, 0.2, 0.1, 100, 100, 10);
        for (int i = 0; i < r.length; i++) {
            double[] val = { x[i] };
            r[i] = obj.eval(val);
        }
        System.out.println(Arrays.toString(r));
        System.out.println(Arrays.toString(obj.err(x, y)));
    }
}

File: python/test/regress_linear.py
#!/usr/bin/env python
from __future__ import print_function
import os, sys

path = os.path.dirname(__file__)
path = os.path.dirname(path)
sys.path.insert(0, path)
if len(sys.argv) > 1:
    sys.stdout = open(sys.argv[1], "w")

import liba  # type: ignore

N = 100
x = liba.new_real([0, 1, 2, 3, 4])
y = liba.new_real([1, 2, 3, 4, 5])

regress = liba.regress_linear([0], 0)

regress.zero()
for i in range(N):
    regress.sgd(x, y, 0.1)
print("y=" + str(regress.coef[0]) + "x+" + str(regress.bias))
print(liba.real_sum(regress.err(x, y)))
print(list(regress.eval(x)))
print(list(regress.err(x, y)))

regress.zero()
for i in range(N):
    regress.bgd(x, y, 0.1)
print("y=" + str(regress.coef[0]) + "x+" + str(regress.bias))
print(liba.real_sum(regress.err(x, y)))
print(list(regress.eval(x)))
print(list(regress.err(x, y)))

regress.zero()
print(regress.mgd(x, y, 0.001, 0.2, 0.1, epoch=N))
print("y=" + str(regress.coef[0]) + "x+" + str(regress.bias))
print(liba.real_sum(regress.err(x, y)))
print(list(regress.eval(x)))
print(list(regress.err(x, y)))

regress.gd(x, 0, 0)

regress.coef = regress.coef
regress.bias = regress.bias
